torch/nn/parallel/distributed_c10d.py

```diff
@@ -278,58 +287,57 @@ def distributed_data_parallel_hook(*unused):
 
             # Current device's bucket is full
             if len(bucket) == self.bucket_sizes[bucket_idx]:
-                self.devs_ready[bucket_idx].append(device_idx)
-                if len(self.devs_ready[bucket_idx]) < len(self.device_ids):
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
                     return
-                #print("bucket: {} is full. bucket size: {}".format(bucket_idx, len(bucket)))
-                self.buckets_ready.append(bucket_idx)
-                if len(self.buckets_to_reduce) > 0 and self.buckets_to_reduce[0] == bucket_idx:
-                    # Reduce the bucket
-                    self._queue_reduction(bucket_idx)
-                    self.buckets_to_reduce.pop(0)
-
-                # If all buckets are ready
-                if len(self.buckets_ready) == len(self.bucket_order):
-                    #print("ready # of buckets: {}".format(self.buckets_ready))
-                    # In case there are unreduced buckets,reduce them all
-                    if len(self.buckets_to_reduce) > 0:
-                        for b_idx in self.buckets_to_reduce:
-                            self._queue_reduction(b_idx)
-                    # A final sync for all the reduction works
-                    self._sync_reduction_works()
+
+                if bucket_idx not in self.buckets_ready:
```

Comment:
How can it not be? Supposedly we only hit this path once per bucket?

Reply:
Safe bet.

Reply:
Can we assert instead? If we hit this twice it must be a logic error.